internal/kojiapi/server.go

```diff
@@ -0,0 +1,244 @@
+	}
+
+	d := server.distros.GetDistro(request.Distribution)
+	if d == nil {
+		http.Error(w, fmt.Sprintf("Unsupported distribution: %s", request.Distribution), http.StatusBadRequest)
+		return
+	}
+
+	type imageRequest struct {
+		manifest distro.Manifest
+		filename string
+	}
+	imageRequests := make([]imageRequest, len(request.ImageRequests))
+
+	for i, ir := range request.ImageRequests {
+		arch, err := d.GetArch(ir.Architecture)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("Unsupported architecture '%s' for distribution '%s'", ir.Architecture, request.Distribution), http.StatusBadRequest)
+			return
+		}
+		imageType, err := arch.GetImageType(ir.ImageType)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("Unsupported image type '%s' for %s/%s", ir.ImageType, ir.Architecture, request.Distribution), http.StatusBadRequest)
+			return
+		}
+		repositories := make([]rpmmd.RepoConfig, len(ir.Repositories))
+		for j, repo := range ir.Repositories {
+			repositories[j].BaseURL = repo.Baseurl
+			repositories[j].GPGKey = repo.Gpgkey
+		}
+		bp := &blueprint.Blueprint{}
+		err = bp.Initialize()
+		if err != nil {
+			panic("Could not initialize empty blueprint.")
+		}
+		packageSpecs, _ := imageType.Packages(*bp)
+		packages, _, err := server.rpmMetadata.Depsolve(packageSpecs, nil, repositories, d.ModulePlatformID(), arch.Name())
+		if err != nil {
+			http.Error(w, fmt.Sprintf("Failed to depsolve base base packages for %s/%s/%s: %s", ir.ImageType, ir.Architecture, request.Distribution, err), http.StatusBadRequest)
+			return
+		}
+		buildPackageSpecs := imageType.BuildPackages()
+		buildPackages, _, err := server.rpmMetadata.Depsolve(buildPackageSpecs, nil, repositories, d.ModulePlatformID(), arch.Name())
+		if err != nil {
+			http.Error(w, fmt.Sprintf("Failed to depsolve build packages for %s/%s/%s: %s", ir.ImageType, ir.Architecture, request.Distribution, err), http.StatusBadRequest)
+			return
+		}
+
+		manifest, err := imageType.Manifest(nil, distro.ImageOptions{Size: imageType.Size(0)}, repositories, packages, buildPackages)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("Failed to get manifest for for %s/%s/%s: %s", ir.ImageType, ir.Architecture, request.Distribution, err), http.StatusBadRequest)
+			return
+		}
+
+		imageRequests[i].manifest = manifest
+		imageRequests[i].filename = imageType.Filename()
+	}
+
+	var ir imageRequest
+	if len(imageRequests) == 1 {
+		// NOTE: the store currently does not support multi-image composes
+		ir = imageRequests[0]
+	} else {
+		http.Error(w, "Only single-image composes are currently supported", http.StatusBadRequest)
+		return
+	}
+
+	// Koji for some reason needs TLS renegotiation enabled.
+	// Clone the default http transport and enable renegotiation.
+	transport := http.DefaultTransport.(*http.Transport).Clone()
+	transport.TLSClientConfig = &tls.Config{
+		Renegotiation: tls.RenegotiateOnceAsClient,
+	}
+
+	k, err := koji.NewFromGSSAPI(request.Koji.Server, &koji.GSSAPICredentials{}, transport)
```

Comment:
I am still not knowledgeable enough about kerberos, but if you don't pass the principal, it will be inferred from the username + default realm, no? So that would become `_osbuild-composer@LOCAL` (if `LOCAL` was the default realm specified in a config file). But in the tests we use `osbuild-krb@LOCAL`. I might be missing something but I don't think this should work.